engine/runtime/src/main/java/org/enso/interpreter/runtime/data/atom/AtomConstructor.java

```diff
@@ -106,49 +135,62 @@ boolean isBuiltin() {
-        scopeBuilder,
-        new ExpressionNode[0],
-        reads,
-        new Annotation[0],
-        args);
+
+    var parts =
+        new InitializationParts(
+            null, LocalScope.empty(), new ExpressionNode[0], reads, new Annotation[0], args);
+    return initializeFields(language, scopeBuilder, CachingSupplier.forValue(parts), fieldNames);
   }
 
   /**
    * Sets the fields of this {@link AtomConstructor} and generates a constructor function.
    *
-   * @param localScope a description of the local scope
-   * @param assignments the expressions that evaluate and assign constructor arguments to local vars
+   * @param language the language implementation
    * @param scopeBuilder the module scope's builder where the accessor should be registered at
-   * @param varReads the expressions that read field values from local vars
+   * @param initializationPartsSupplier the function supplying the parts required for initialization
+   * @param argumentsLength the number of the arguments
    * @return {@code this}, for convenience
    */
   public AtomConstructor initializeFields(
       EnsoLanguage language,
-      SourceSection section,
-      LocalScope localScope,
       ModuleScope.Builder scopeBuilder,
-      ExpressionNode[] assignments,
-      ExpressionNode[] varReads,
-      Annotation[] annotations,
-      ArgumentDefinition... args) {
+      Supplier<InitializationParts> initializationPartsSupplier,
+      String[] fieldNames) {
     CompilerDirectives.transferToInterpreterAndInvalidate();
-    assert boxedLayout == null : "Don't initialize twice: " + this.name;
-    if (args.length == 0) {
+    assert accessor == null : "Don't initialize twice: " + this.name;
+    this.fieldNames = fieldNames;
+    if (fieldNames.length == 0) {
       cachedInstance = BoxingAtom.singleton(this);
     } else {
       cachedInstance = null;
     }
-    boxedLayout = Layout.createBoxed(args);
+    CachingSupplier<InitializationResult> initializationResultSupplier =
+        CachingSupplier.wrap(
+            () -> {
+              var parts = initializationPartsSupplier.get();
+              var constructorFunction =
+                  buildConstructorFunction(
+                      language,
+                      parts.section,
+                      parts.localScope,
+                      scopeBuilder,
+                      parts.assignments,
+                      parts.varReads,
+                      parts.annotations,
+                      parts.args);
+              var layout = Layout.createBoxed(parts.args);
+              return new InitializationResult(constructorFunction, layout);
+            });
+    this.boxedLayout =
+        initializationResultSupplier.map(initializationResult -> initializationResult.layout);
     this.constructorFunction =
-        buildConstructorFunction(
-            language, section, localScope, scopeBuilder, assignments, varReads, annotations, args);
+        initializationResultSupplier.map(
```

Comment:
Interesting use of the new `map` function. The alternative would be to have a single field `Supplier<InitializationResult>` stored in the `AtomConstructor`. I seem to like it a bit more - shows the _"fixed"_ vs. the lazily computed part of the structure.